src/state.rs

```diff
@@ -503,37 +505,44 @@ impl ProgressDrawTarget {
 
     /// Apply the given draw state (draws it).
     pub(crate) fn apply_draw_state(&mut self, draw_state: ProgressDrawState) -> io::Result<()> {
-        let (term, last_line_count, last_draw) = match self.kind {
+        match self.kind {
             ProgressDrawTargetKind::Term {
                 ref term,
                 ref mut last_line_count,
                 rate,
                 ref mut last_draw,
-            } if draw_state.finished || draw_state.force_draw || last_draw.elapsed() > rate => {
-                (term, last_line_count, last_draw)
+                ref mut leaky_bucket,
+            } => {
+                if draw_state.finished || draw_state.force_draw || *leaky_bucket < MAX_GROUP_SIZE {
+                    if !draw_state.lines.is_empty() && draw_state.move_cursor {
+                        term.move_cursor_up(*last_line_count)?;
+                    } else {
+                        term.clear_last_lines(*last_line_count)?;
+                    }
+            
+                    draw_state.draw_to_term(term)?;
+                    term.flush()?;
+                    *last_line_count = draw_state.lines.len() - draw_state.orphan_lines;
+                    *leaky_bucket += 1.0;
+                }
+                let ticks = last_draw.elapsed().as_secs_f64()/rate.as_secs_f64();
+                *leaky_bucket -= ticks;
+                if *leaky_bucket < 0.0 {
+                    *leaky_bucket = 0.0;
+                }
+                *last_draw = Instant::now();
+                Ok(())
```

Comment:
Can we keep this outside the match, as before? I prefer that stylistically, and it makes this change easier to review.

Reply:
We can, and I am happy to do that if it is what you want, I moved it inside as it was the only branch that the code applied to and it needs all the fields of `ProgressDrawTargetKind::Term` so it felt cleaner to move it inside that pass them all around in a tuple.

It would now look like:
`let (term, last_line_count, rate, last_draw, leaky_bucket) = match self.kind {`

If you confirm that you prefer it that way I will do that.

Reply:
I can split it into two commits if that makes it nicer

Reply:
Maybe it would be an improvement to wrap all those fields in a separate type instead (in a separate commit)?

Reply:
It looks like you made the opposite change from that I intended. I would like still like to keep the core drawing routine out of the match statement.

Reply:
Sorry I had pulled it out this way before I saw your comment, I'll do it the way you requested now.

Reply:
I have moved the leaky bucket logic to it's own struct, and put the rest of the code back at the end.